src/mlu/user_data_proxy.rs

```diff
@@ -0,0 +1,45 @@
+use std::marker::PhantomData;
+
+use mlua::{AnyUserData, Error, Lua, ToLua, UserData};
+
+use crate::{TypeBody, TypeName};
+
+/// A userdata which can be used as a static proxy
+pub trait StaticUserdata: UserData + 'static {}
+impl<T: UserData + 'static> StaticUserdata for T {}
+
+/// A newtype storing userdata created via [`mlua::Lua::create_proxy`].
+///
+/// if `T` implements TypeName or TypeBody the implementations are forwarded to this type.
+pub struct UserDataProxy<'lua, T: StaticUserdata> {
+    user_data: AnyUserData<'lua>,
+    ph_: PhantomData<T>,
+}
+
+impl<'lua, T: StaticUserdata> UserDataProxy<'lua, T> {
+    /// Creates a new UserDataProxy
+    pub fn new(lua: &'lua Lua) -> Result<Self, Error> {
+        Ok(Self {
+            user_data: lua.create_proxy::<T>()?,
+            ph_: Default::default(),
+        })
+    }
+}
+
+impl<T: StaticUserdata + TypeName> TypeName for UserDataProxy<'_, T> {
+    fn get_type_parts() -> std::borrow::Cow<'static, [crate::NamePart]> {
+        T::get_type_parts()
```

Comment:
this would make the UserDataProxy be the same type as the thing it is proxying. Wouldn't it be better to give it some prefix/postfix to distinguish between the 2?

Reply:
I thought it wouldn't make sense to do that, the entire point of the proxy is to have the same metatable as the proxied type.

The way I wanted to use it (and imagine it being used) is since static methods defined on the original type will pop up on that type anyway, there's no point in duplicating the documentation for the type for "only static" parts.

I.e. having this proxy removes the need for a second "global" version of the type in the documentation whatsover.

Also If I add a prefix, the example no longer works since the d.tl file will be referencing a non existing type (i.e. it would have to be registered as its own type, which sort of defeats the purpose)

Reply:
If they share the same typename then the proxy (as far as any statically typed language is concerned) can be given to any function that expects the real type while to my understanding that is actually not possible.

I also think it is less confusing to have the 2 be totally separate types as then it is very clear from the documentation that one doesn't have all the methods that the other has. It also makes the exposed global of said proxy more clear `global Vec2Class : Vec2` vs `global Vec2Class : Vec2Class`. To me, the first one sounds like there is a proper Vec2 named "Vec2Class" which makes me curious what is actually stored in there. While the second makes it clear that it is not a normal Vec2.

You can of course get around that by adding documentation but.... now you add documentation for something that can also be done in a way that languages understand. So, that doesn't sound that great to me to be honest. 

Ideally, the real type would extend the proxy. That way, the static methods only have to get documented once. However tealr has no way to express that and it sounds a bit out of scope for this PR to add that (though, I am happy to merge that as part of this).

Reply:
Hmm, If someone wants to have a completely separate global type, then they are free to just go the old way, and to create and export a completely new UserData type, this PR wouldn't stop them from doing that,
This simply adds another parallel way of documenting globals which requires much less writing, and less metatables (and is likely more efficient all the way around).

But anyhow how would you like me to proceed with the PR then? It sounds like you want `process_type::<MyType>` to generate not one but, two type entries, one for the proxy and one for the main type, and then this wrapper would be referencing that, or are you happy with how things are in this PR right now ?

Reply:
I think separating it into 2 is best. Then later on it can easily be changed so the real type extends the proxy version.

I rather have the documentation and definition files reflect what is really going on at the cost of it being more verbose with a clear way to solve that (adding the ability for types to inherit each other) versus having them hide how things are working. I find this especially important for definition files as they tell other languages what is and what is not allowed. Lying to those languages for the purpose of making it less verbose, while the user is most likely not even going to look at them and is instead expected to look at the generated html documentation sounds weird to me.

Once tealr_doc_gen accepts custom templates then it should also be possible for people to remove any mention of the Proxies or deal with them in some other way. This can be done even without setting up inheritance for as long as the Proxies have a consistent pattern that is easy to check for. Removing the proxy on tealr's end however would make custom handling of them harder.

Reply:
How would this work API wise? Are we to assume all "functions" are static ? But then that won't hold if a function accepts `Self` and tries to get state out of it.

One way is to have something like:
```rust
    fn add_methods<'lua, T: TealDataMethods<'lua, Self>>(methods: &mut T, class_methods: &mut T)
```
and let the user decide which functions end up on the class, or maybe even use a default impl like:
```rust
    fn add_class_methods<'lua, T: TealDataMethods<'lua, Self>>(methods: &mut T){
        Ok(false)
    }
```
to not change the current API. This would of course duplicate some of the machinery, or at least require a flag on each method, so that we know which methods to put on the "Class" Type.

This would probably require changes to `TypeBody` to:
```rust
    fn get_type_body() -> TypeGenerator;
    fn get_class_type_body() -> Option<TypeGenerator>;
```

What do you think? I am happy to take this on

Reply:
> Are we to assume all "functions" are static ? But then that won't hold if a function accepts `Self` and tries to get state out of it.

The `Self` would still refer to the unproxied type, right? Which then once asked for the TypeName will still give back the correct typename rather than the proxied type name? Or am I missing something?

```rs
#[derive(Clone, Copy)]
struct Example(u32);

impl TealData for Example {
    fn add_methods<'lua, T: TealDataMethods<'lua, Self>>(methods: &mut T) {
        methods.add_function_mut("example_function_mut", |_, x: Self| {
            let z = x.0;
            x.0 = 0;
           Ok(x)
        })
    }
}
```
Here, the `Self` would refer to `Example` right? So, it would become a `example_function_mut: function(Example):Example` regardless if the method gets added to a proxy object or to the real one? 

Or.. am I missing something (obvious?)

Ideally I'd like to stay as close to mlua's method of adding class methods as possible. Which would assume that every function/mut_function is a static. If that can't be done my second concern is to keep `TypeBody` and the other common parts as free of mlua/rlua specific things as possible. I think it is best if the common parts stay as free on what a scripting language api should have as possible. If doing so ends up repeating some machinery then so be it.

Reply:
No yeah as I said in my EDIT, the types of the proxy and the `Example` are different so everything works, I am working on a quick draft of this functionality, I don't think any API changes are necessary actually.

Btw I think tealr could benefit from a linear string writer akin to this: https://github.com/makspll/bevy_mod_scripting/blob/add_general_api/bevy_api_gen/src/writer.rs

There are a lot of big string operations which could likely be "flattened" out and could make the code much more readable!

Reply:
yea, the plan is to move `.d.tl` file generation over to `tealr_doc_gen` and use a simple etlua template to generate it. That should end up making it easier for people to create definition files for more languages than just teal. Though, tealr still has some ways to go to make that easy even after that change due to how the type information is stored is basically teal's own syntax.